accepted/system-range/system-range.md

```diff
@@ -0,0 +1,246 @@
+
+foreach (var value in Range.Create(1, 4))
+    Console.WriteLine(value);
+```
+
+With C# syntax this will look as follows:
+
+```C#
+var fiveToTen = 5..11; // Equivalent to Range.Create(5, 11)
+```
+
+### Unbounded ranges
+
+```C#
+// Ranges can be unbounded:
+
+var fiveToEnd = 5..;  // Equivalent to Range.FromStart(5) i.e. missing upper bound
+var startToTen = ..1; // Equivalent to Range.ToEnd(1) i.e. missing lower bound
+var everything = ..;  // Equivalent to Range.All() i.e. missing upper and lower bound
+```
+
+## Requirements
+
+### Goals
+
+* We need a type that both, plays well with existing concepts in the base class
+  library, as well as the intended syntax for C#
+* We need to ensure we can extend the notion of ranges in the base class library
+  in the future (for example, by supporting element types other than `Int32`).
+
+### Non-Goals
+
+* Supporting other numeric types or even offer a generic range type
+* Creating overloads on existing types that will accept `Range` (that doesn't
+  mean we shouldn't be doing it, but that it will be covered by a separate
+  design proposal).
+* Supporting `Index` and `Range` on multi-dimensional arrays and non-zero based
+  arrays (e.g. instances of `System.Array`).
+    - This work needs more thought and is currently scoped out.
+
+## Design
+
+### Proposed API
+
+```C#
+namespace System
+{
+    public readonly struct Index
+    {
+        private readonly int _value;
+
+        public int Value => _value < 0 ? ~_value : _value;
+        public bool FromEnd => _value < 0;
+
+        public Index(int value, bool fromEnd = false)
+        {
+            if (value < 0)
+                throw new ArgumentException("Index must not be negative.", nameof(value));
+
+            _value = fromEnd ? ~value : value;
+        }
+
+        public static implicit operator Index(int value) => new Index(value);
+    }
+
+    public readonly struct Range
+    {
+        public Index Start { get; }
+        public Index End { get; }
+
+        private Range(Index start, Index end)
+        {
+            Start = start;
+            End = end;
+        }
+
+        public static Range Create(Index start, Index end) => new Range(start, end);
+        public static Range FromStart(Index start) => new Range(start, new Index(0, fromEnd: true));
+        public static Range ToEnd(Index end) => new Range(new Index(0, fromEnd: false), end);
+        public static Range All() => new Range(new Index(0, fromEnd: false), new Index(0, fromEnd: true));
+    }
+```
+
+### Indexing with SZ-arrays
+
+SZ-arrays (i.e. `T[]` in C#) do not have an API that represents the indexer.
+Instead, they are specialized as IL opcodes to load and store elements in such
+an array. The indexer is part of the language semantics and thus doesn't have an
+API representation.
+
+For indexing using `Index` the compiler doesn't have to do much, thus the
+handling is done directly by the compiler, i.e. this code:
+
+```C#
+T[] array = GetArray(;)
```

Comment:
`GetArray();` – same below in line 148

Reply:
It's a new language feature. It makes sure that the calling method doesn't allocate. In this case though, the method has to allocate so I changed it back to `GetArray();`.

Reply:
:trollface: